### PR TITLE
fix(platform-ios): detect Podfile and .xcworkspace

### LINKED ISSUE #1435: run-ios bug: Try/catch removed from `warnAboutManuallyLinkedLibs.ts` causing exceptions to propagate/crash
## Environment

```
System:
  OS: macOS 11.4
  CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  Memory: 921.95 MB / 16.00 GB
  Shell: 3.2.57 - /bin/bash
Binaries:
  Node: 16.4.0 - /usr/local/bin/node
  Yarn: 1.22.10 - /usr/local/bin/yarn
  npm: 7.18.1 - /usr/local/bin/npm
  Watchman: 4.9.0 - /usr/local/bin/watchman
Managers:
  CocoaPods: 1.10.1 - /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
  Android SDK:
    API Levels: 28, 29, 30
    Build Tools: 28.0.3, 29.0.3, 30.0.1
    System Images: android-30 | Google APIs Intel x86 Atom
    Android NDK: Not Found
IDEs:
  Android Studio: 4.1.2 4.1.2
  Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
Languages:
  Java: 11.0.1 - /usr/bin/javac
npmPackages:
  @react-native-community/cli: Not Found
  react: 17.0.1 => 17.0.1 
  react-native: 0.64.1 => 0.64.1 
  react-native-macos: ^0.63.0 => 0.63.36 
npmGlobalPackages:
  *react-native*: Not Found
```

## Description

This refers to https://github.com/microsoft/react-native-test-app/issues/375

I was working on moving a community module to use [React Native Test App](https://github.com/microsoft/react-native-test-app/) to manage the test app of a community module (https://github.com/react-native-menu/menu), and ran into an error running `run-ios`. After investigation, it seems the issue happened upstream beginning in https://github.com/react-native-community/cli/pull/1270 .

A try-catch was removed from `packages/platform-ios/src/link/warnAboutManuallyLinkedLibs.ts`, causing the exception to bubble up and crash. This seems like an error, because an Android fix shouldn't have broken something downstream for iOS.

## Reproducible Demo

A little bit contrived since I reproed this using a downstream module, but:
```
git clone git@github.com:react-native-menu/menu.git
cd menu
mv example/ example2/ # This is to just move the existing test app out of the way temporarily
yarn add react-native-test-app --dev
yarn init-test-app
# The options I chose
# ✔ What is the name of your test app? … MenuExample
# ✔ Which platforms do you need test apps for? › Android, iOS
# ✔ Where should we create the new project? … example
yarn
yarn build:ios
pod install --project-directory=ios
yarn ios
```

### PR DESCRIPTION
**Pull request by tido64**
Summary:
---------

`react-native config` fails to recognize projects that contain only a Podfile or `.xcworkspace`, e.g. because the Xcode project is generated. This change makes it output a config if either are found.

Resolves #1435
Resolves the iOS part of #1054

Test Plan:
----------

```
git clone git@github.com:react-native-menu/menu.git
cd menu
mv example/ example2/ # This is to just move the existing test app out of the way temporarily
yarn add react-native-test-app --dev
yarn init-test-app
# The options I chose
# ✔ What is the name of your test app? … MenuExample
# ✔ Which platforms do you need test apps for? › Android, iOS
# ✔ Where should we create the new project? … example

cd example
yarn
yarn link "@react-native-community/cli"
yarn link "@react-native-community/cli-platform-ios"
pod install --project-directory=ios
yarn ios
```

### CODE REVIEW

**Review by thymikee — APPROVED** [review]
LGTM. What do you think @grabbou?

**Comment by tido64** [comment]
@grabbou, @Esemesek: Gentle ping. Let me know if you have any comments and I'll try to address them. 😄

**Review by grabbou — CHANGES_REQUESTED** [review]
Thanks for the PR. I left a series of comments.

Before going to addressing them, I wanted to better understand the nature of this PR. 

In what scenario, an `xcworkspace` will be present, but `xcodeproj` not?

Why should we allow CLI to detect projects that don't have `.xcworkspace` and `.xcodeproj`, but have a `Podfile` only? CLI is designed to support projects, not dependencies.

File: packages/platform-ios/src/config/index.ts [hunk 1]
```diff
@@ -38,31 +39,41 @@ export function projectConfig(folder: string, userConfig: IOSProjectParams) {
...
   return {
     sourceDir,
     folder,
-    pbxprojPath: path.join(projectPath, 'project.pbxproj'),
-    podfile: findPodfilePath(projectPath),
+    pbxprojPath:
+      projectPath && isXcodeProject(projectPath)
+        ? path.join(projectPath, 'project.pbxproj')
+        : null,
+    podfile: podfile && path.resolve(podfile),
```
[the remark this answers is not in the document]
Comment (thymikee) [reply packages/platform-ios/src/config/index.ts:68]: Hm right, with `!project && !podfile` guard we don't need to check either of those

**Comment by grabbou** [comment]
@tido64 how about we hop on a call at some point this or next week to discuss this?

**Comment by tido64** [comment]
> In what scenario, an `xcworkspace` will be present, but `xcodeproj` not?
> 
> Why should we allow CLI to detect projects that don't have `.xcworkspace` and `.xcodeproj`, but have a `Podfile` only? CLI is designed to support projects, not dependencies.

I filed a feature request on this last year or so here: https://github.com/react-native-community/cli/issues/1054.

Technically, only `Podfile` and `.xcworkspace` need to reside in the same folder. `.xcworkspace` can point to a `.xcodeproj` anywhere in the file system and does not have the same restriction. Moreover, if you're using a project generator, you may not find a `.xcworkspace` there at all.

A more specific example are projects using [`react-native-test-app`](https://github.com/microsoft/react-native-test-app) to generate both `.xcworkspace` and `.xcodeproj` during `pod install` (there was an [RFC](https://github.com/react-native-community/discussions-and-proposals/pull/204) some time back if you're interested in details). Such projects may only have a `Podfile` in the `ios` folder ([example](https://github.com/react-native-async-storage/async-storage/tree/master/example/ios), [example 2](https://github.com/react-native-menu/menu/pull/142)). The generated `.xcodeproj` is hidden since it's managed by `react-native-test-app` and is not meant to be checked in. This is to hide away a piece of complexity that users need not care about. This is especially beneficial when they need to upgrade (or downgrade) `react-native`.

**Comment by tido64** [comment]
> @tido64 how about we hop on a call at some point this or next week to discuss this?

Yes, we can do that as well. I know you're super busy so let me know when you have time?

**Comment by grabbou** [comment]
I will reach out to you directly over Discord and we will send a note when we're done

**Pull request closed without merging**